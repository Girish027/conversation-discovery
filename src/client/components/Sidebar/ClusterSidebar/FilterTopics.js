import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Map, fromJS } from 'immutable';
import { connect } from 'react-redux';
import _ from 'lodash';

import {
  clusterSelectors,
  clusterActions,
  initialFilterState,
} from '../../../state/clusterState';
import { runSelectors } from '../../../state/runsState';
import { Checkbox, DropDown, TextField } from '@tfs/ui-components';
import { filterPopupStyle } from '../../../styles';
import 'styles/sidebar.scss';

export class FilterTopics extends Component {
  constructor(props) {
    super(props);
    this.props = props;

    this.toggleCheckbox = this.toggleCheckbox.bind(this);
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleDropDownChange = this.handleDropDownChange.bind(this);

    this.volumeFilterBy = {
      count: 'Count',
      percent: 'Percent'
    };

    this.state = this.props.filters.toJS();
    this.flag = {
      volume: this.state.flag.volume,
      similarityCutOff: this.state.flag.similarityCutOff
    };
    this.filterItems = {
      finalized: 'finalized',
      open: 'open',
      similarityCutOff: 'similarityCutOff',
      volume: 'volume'
    };
  }

  handleDropDownChange(selectedValue) {
    const { dispatch } = this.props;
    if (selectedValue === this.volumeFilterBy.percent) {
      if (!this.flag.volume) {
        this.setState({
          selectedTopFilter: this.volumeFilterBy.percent,
          volume: {},
        });
      } else {
        this.setState({
          selectedTopFilter: this.volumeFilterBy.percent,
        }, () => {
          let newFilters = {};
          if (this.flag.volume) {
            if (!_.isEmpty(this.state.volume)) {
              newFilters = {
                volume: { volumeByPercent: this.state.volume.volumeByCount || this.state.volume.volumeByPercent },
                flag: this.flag,
                selectedTopFilter: this.volumeFilterBy.percent
              };
              dispatch(clusterActions.setFilter(newFilters));
            } else {
              newFilters = {
                volume: {},
                flag: this.flag,
                selectedTopFilter: this.volumeFilterBy.percent
              };
              dispatch(clusterActions.setFilter(newFilters));
            }
          }
        });
      }
    } else if (selectedValue === this.volumeFilterBy.count) {
      if (!this.flag.volume) {
        this.setState({
          selectedTopFilter: this.volumeFilterBy.count,
          volume: {},
        });
      } else {
        this.setState({
          selectedTopFilter: this.volumeFilterBy.count,
        }, () => {
          let newFilters = {};
          if (this.flag.volume) {
            if (!_.isEmpty(this.state.volume)) {
              newFilters = {
                volume: { volumeByCount: this.state.volume.volumeByPercent || this.state.volume.volumeByCount },
                flag: this.flag,
                selectedTopFilter: this.volumeFilterBy.count
              };
              dispatch(clusterActions.setFilter(newFilters));
            } else {
              newFilters = {
                volume: {},
                flag: this.flag,
                selectedTopFilter: this.volumeFilterBy.count
              };
              dispatch(clusterActions.setFilter(newFilters));
            }
          }
        });
      }
    }
  }

  handleInputChange(val, filteredItem) {
    const { selectedTopFilter } = this.state;
    const { dispatch } = this.props;

    if (filteredItem === this.filterItems.similarityCutOff) {
      this.setState({
        similarityCutOff: parseFloat(val),
      }, () => {
        const { similarityCutOff } = this.state;
        let newFilters = {};
        if (this.flag.similarityCutOff) {
          if (similarityCutOff) {
            newFilters = {
              similarityCutOff: this.state.similarityCutOff,
              flag: this.flag
            };
            dispatch(clusterActions.setFilter(newFilters));
          }
        } else {
          newFilters = {
            similarityCutOff: '',
            flag: this.flag
          };
          dispatch(clusterActions.setFilter(newFilters));
        }
      });
    } else if (filteredItem === this.filterItems.volume && selectedTopFilter) {
      if (selectedTopFilter === this.volumeFilterBy.count) {
        this.setState({
          volume: {
            volumeByCount: parseInt(val, 10),
          }
        }, () => {
          let newFilters = {};
          if (this.flag.volume) {
            if (!_.isEmpty(this.state.volume)) {
              newFilters = {
                volume: this.state.volume,
                flag: this.flag,
                selectedTopFilter
              };
            }
            dispatch(clusterActions.setFilter(newFilters));
          } else {
            newFilters = {
              volume: {},
              flag: this.flag,
              selectedTopFilter
            };
            dispatch(clusterActions.setFilter(newFilters));
          }
        });
      } else if (selectedTopFilter === this.volumeFilterBy.percent) {
        this.setState({
          volume: {
            volumeByPercent: parseInt(val, 10),
          }
        }, () => {
          let newFilters = {};
          if (this.flag.volume) {
            if (!_.isEmpty(this.state.volume)) {
              newFilters = {
                volume: this.state.volume,
                flag: this.flag,
                selectedTopFilter
              };
            }
            dispatch(clusterActions.setFilter(newFilters));
          } else {
            newFilters = {
              volume: {},
              flag: this.flag,
              selectedTopFilter
            };
            dispatch(clusterActions.setFilter(newFilters));
          }
        });
      }
    }
  }

  toggleCheckbox(checked, value) {
    const { dispatch } = this.props;
    const { volume, selectedTopFilter } = this.state;
    const { filters: filtersImmutable } = this.props;
    const filters = filtersImmutable.toJS();

    /* TODO: revisit once api for similarity score and top volume is ready, currently this is dummy code to check the filtering */

    if (value === this.filterItems.volume) {
      this.flag.volume = checked;
      let newFilters = {};
      if (checked) {
        if (!_.isEmpty(volume)) {
          newFilters = {
            volume,
            flag: this.flag,
            selectedTopFilter
          };
        }
        dispatch(clusterActions.setFilter(newFilters));
      } else {
        newFilters = {
          volume: {},
          flag: this.flag,
          selectedTopFilter
        };
        dispatch(clusterActions.setFilter(newFilters));
      }
    }

    if (value === this.filterItems.similarityCutOff) {
      this.flag.similarityCutOff = checked;
      let newFilters = {};
      if (checked) {
        if (this.state.similarityCutOff) {
          newFilters = {
            similarityCutOff: this.state.similarityCutOff,
            flag: this.flag
          };
          dispatch(clusterActions.setFilter(newFilters));
        }
      } else {
        newFilters = {
          similarityCutOff: '',
          flag: this.flag
        };
        dispatch(clusterActions.setFilter(newFilters));
      }
    }

    if (value === this.filterItems.finalized) {
      const newFilters = {
        finalized: !filters.finalized,
      };
      dispatch(clusterActions.setFilter(newFilters));
    }
    if (value === this.filterItems.open) {
      const newFilters = {
        open: !filters.open,
      };
      dispatch(clusterActions.setFilter(newFilters));
    }
  }

  render() {
    const { similarityCutOff, volume } = this.state;
    const { filters: filtersImmutable } = this.props;
    const filters = filtersImmutable.toJS();
    let textFieldVal = '';

    if (!_.isEmpty(filters.volume)) {
      textFieldVal = filters.volume.volumeByCount ? filters.volume.volumeByCount : filters.volume.volumeByPercent;
    } else {
      textFieldVal = !_.isEmpty(volume) && (volume.volumeByPercent ? volume.volumeByPercent : volume.volumeByCount);
    }

    return (
      <div>
        <div style={filterPopupStyle.filterTopContainer}>
          <Checkbox
            key='one'
            checked={!_.isEmpty(filters.volume) && (filters.volume.volumeByCount !== null || filters.volume.volumeByPercent !== null)}
            label='Top'
            onChange={(event) => {
              this.toggleCheckbox(event.target.checked, 'volume');
            }}
          />

          <TextField
            type='number'
            min={0}
            max={100}
            class='filter-input'
            defaultValue={textFieldVal}
            onChange={
              (event) => this.handleInputChange(event.target.value, this.filterItems.volume)
            }
            name='top-topics'
            placeholder={50}
            styleOverride={filterPopupStyle.textboxStyle}
          />

          <DropDown
            itemList={[this.volumeFilterBy.count, this.volumeFilterBy.percent]}
            labelName=' '
            selectedIndex={filters.selectedTopFilter === 'Percent' ? 1 : 0}
            styleOverride={filterPopupStyle.dropdownStyle}
            shouldRetainState
            onItemSelected={(selectedValue) => this.handleDropDownChange(selectedValue)}
          />
        </div>

        <div style={filterPopupStyle.filterSimilarity}>
          <Checkbox
            key='two'
            checked={filters.similarityCutOff !== '' || this.flag.similarityCutOff === true}
            value={this.filterItems.similarityCutOff}
            disabled={similarityCutOff === '' && !(similarityCutOff === '' && filters.similarityCutOff)}
            label='Similarity Score greater than'
            onChange={(event) => {
              this.toggleCheckbox(event.target.checked, this.filterItems.similarityCutOff);
            }}
          />
          <TextField
            type='number'
            min={0}
            max={100}
            class='filter-input'
            styleOverride={filterPopupStyle.textboxStyle}
            placeholder={0.08}
            onChange={(event) => this.handleInputChange(event.target.value, this.filterItems.similarityCutOff)}
            name='similarity-score'
            defaultValue={filters.similarityCutOff !== '' ? filters.similarityCutOff : similarityCutOff}
          />
        </div>

        <Checkbox
          key='three'
          checked={filters.finalized}
          value={this.filterItems.finalized}
          label='Finalized Topics'
          styleOverride={{ paddingBottom: '5px' }}
          onChange={(event) => {
            this.toggleCheckbox(event.target.checked, this.filterItems.finalized);
          }}
        />

        <Checkbox
          key='four'
          checked={filters.open}
          value={this.filterItems.open}
          label='Open Topics'
          onChange={(event) => {
            this.toggleCheckbox(event.target.checked, this.filterItems.open);
          }}
        />
      </div>
    );
  }
}

FilterTopics.propTypes = {
  filters: PropTypes.object,
  dispatch: PropTypes.func,
};

FilterTopics.defaultProps = {
  filters: fromJS(initialFilterState),
  dispatch: () => {},
};

const mapStateToProps = (state) => {
  const allFilters = clusterSelectors.getFilters(state) || new Map({});
  const selectedRunId = runSelectors.getSelectedRunId(state) || '';
  const filters = allFilters.get(selectedRunId);
  return {
    filters,
  };
};

const mapDispatchToProps = (dispatch) => ({
  dispatch,
});

export default connect(mapStateToProps, mapDispatchToProps)(FilterTopics);
