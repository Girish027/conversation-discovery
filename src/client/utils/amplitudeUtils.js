// import amplitude from 'amplitude-js/amplitude';

export const initializeAnalytics = (analyticsKey, userName = '') => {
  // const analytics = amplitude.getInstance();
  // analytics.init(analyticsKey, userName);
};

export const logAmplitudeEvent = (eventName, eventSpecificData = {}) => {
  // amplitude.getInstance().logEvent(eventName, {
  //   ...eventSpecificData,
  // });
};
