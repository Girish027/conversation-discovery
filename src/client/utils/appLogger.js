// log to console if method level meets log level
function appLogger(options) {
  // indexes of method names match values of levels properties
  const loggerMethods = ['trace', 'debug', 'info', 'warn', 'error'];
  const levels = {
    TRACE: 0,
    DEBUG: 1,
    INFO: 2,
    WARN: 3,
    ERROR: 4,
    SILENT: 5
  };
  const envLevels = {
    development: levels.TRACE,
    production: levels.SILENT,
    test: levels.ERROR
  };

  // environment variables can override default log level
  let logLevel = levels.INFO;
  if (typeof process.env.LOG_LEVEL !== 'undefined') {
    logLevel = levels[process.env.LOG_LEVEL];
  } else if (typeof process.env.NODE_ENV !== 'undefined') {
    logLevel = envLevels[process.env.NODE_ENV];
  }

  const logger = {};
  loggerMethods.forEach((methodName, methodLevel) => {
    logger[methodName] = (...args) => {
      // only print out if method level is not less than log level
      if (methodLevel >= logLevel) {
        let msgs = [];
        if (options && options.tag) {
          msgs.push(options.tag);
        }
        msgs = msgs.concat(args);
        /* eslint-disable no-alert, no-console */
        console.log(...msgs);
      }
    };
  });
  return logger;
}

export default appLogger;
