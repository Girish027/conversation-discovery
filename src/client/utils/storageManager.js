import Constants from 'components/../constants';

const loadCFDStore = () => {
  let savedStore = localStorage.getItem(Constants.localStorageBucket);
  if (savedStore) {
    try {
      savedStore = JSON.parse(savedStore);
    } catch (err) {
      savedStore = {};
    }
  } else {
    savedStore = {};
  }
  return savedStore;
};

export const loadStorage = (key, defaultReturnValue = null) => {
  const savedStore = loadCFDStore();
  const { [key]: value = defaultReturnValue } = savedStore;
  return value;
};

export const saveStorage = (key, value) => {
  try {
    const savedStore = loadCFDStore();
    savedStore[key] = value;
    localStorage.setItem(Constants.localStorageBucket, JSON.stringify(savedStore));
  } catch (err) {
    /* DO nothing */
  }
};

export const clearStorage = (key) => {
  try {
    const { [key]: removedData, ...savedStore } = loadCFDStore();
    localStorage.setItem(Constants.localStorageBucket, JSON.stringify(savedStore));
  } catch (err) {
    /* DO nothing */
  }
};
