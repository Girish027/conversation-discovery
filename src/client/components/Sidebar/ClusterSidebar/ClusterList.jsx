import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
// import _ from 'lodash';
import { clusterListStyle } from '../../../styles';
import '../../../styles/collapsible.scss';
import {
  clusterSelectors,
  clusterActions,
} from 'state/clusterState';
import { getPathnameSelector } from 'state/routerState';
import { Button, TableSorter } from '@tfs/ui-components';
import constants from '../../../constants';
import { prepareClusterName } from '../../../utils/stringOperations';

export class ClusterList extends PureComponent {
  constructor(props) {
    super(props);
    this.props = props;
    this.state = {
      isHover: false,
      focusedCluster: '',
    };

    this.onClickVolume = this.onClickVolume.bind(this);
    this.toggleFocus = this.toggleFocus.bind(this);
  }

    onSelectCluster = (clusterId) => {
      const { dispatch } = this.props;
      dispatch(clusterActions.selectCluster(clusterId));
    };

    onClickVolume() {
      const { handleSorting } = this.props;
      handleSorting();
    }

    toggleFocus = (clusterId) => {
      const { isHover } = this.state;
      this.setState({
        isHover: !isHover,
        focusedCluster: clusterId
      });
    }

    getLabelStyle = (clusterId) => {
      const { selectedClusterId } = this.props;
      const { isHover, focusedCluster } = this.state;
      let clusterStyle = {};

      if (selectedClusterId === clusterId) {
        clusterStyle = clusterListStyle.selectedItem;
      } else if (isHover && focusedCluster === clusterId) {
        clusterStyle = clusterListStyle.innerListStyleHover;
      } else {
        clusterStyle = clusterListStyle.innerListStyle;
      }
      return clusterStyle;
    }

    clusterConfig = (clustersList, sortOrder) => {
      if (sortOrder === constants.sortOrder.asc) {
        clustersList.sort((a, b) => a.count - b.count);
      } else if (sortOrder === constants.sortOrder.desc) {
        clustersList.sort((a, b) => b.count - a.count);
      } else {
        clustersList.sort((a, b) => b.count - a.count);
        clustersList.push(...clustersList.splice(clustersList.findIndex((v) => v.topic === constants.OTHER_CLUSTER), 1));
      }
    }

    render() {
      const { sortOrder } = this.props;
      let { clusters: clustersList } = this.props;
      if ((Array.isArray(clustersList)) && clustersList.length > 0) {
        const arr = [];
        clustersList.forEach((element) => {
          const arrays = element.clusters;
          arrays.forEach((el) => {
            const obj = {};
            obj.topic = el.clusterName;
            obj.count = el.count;
            obj.clusterId = el.clusterId;
            arr.push(obj);
          });
        });
        clustersList = arr;
      }
      this.clusterConfig(clustersList, sortOrder);
      return (
        <div>
          <div style={clusterListStyle.headerStyle}>
            <span style={clusterListStyle.headerTopicStyle}>Potential Intents</span>
            <div style={clusterListStyle.seperatorStyle}>
              <Button
                name='volume-btn'
                onClick={this.onClickVolume}
                type='flat'
                styleOverride={clusterListStyle.headerButtonStyle}
              >
                Volume
              </Button>
              <TableSorter sortOrder={sortOrder} />
            </div>
          </div>
          <hr style={clusterListStyle.lineStyle} />
          <ul style={clusterListStyle.ulStyle}>
            {clustersList.map((cluster) => (
              <li key={cluster.clusterId}>
                <div
                  role='button'
                  tabIndex={0}
                  onKeyDown={() => this.onSelectCluster(cluster.clusterId)}
                  onClick={() => this.onSelectCluster(cluster.clusterId)}
                  style={this.getLabelStyle(cluster.clusterId)}
                  onMouseOver={() => this.toggleFocus(cluster.clusterId)}
                  onMouseOut={() => this.toggleFocus(cluster.clusterId)}
                  onFocus={() => {}}
                  onBlur={() => {}}
                >
                  <div style={clusterListStyle.innerLabel}>
                    {prepareClusterName(cluster.topic)}
                  </div>
                  <div style={clusterListStyle.innerSeperatorStyle}>
                    {cluster.count}
                  </div>
                </div>
              </li>
            ))}
          </ul>
        </div>
      );
    }
}

ClusterList.propTypes = {
  dispatch: PropTypes.func,
  selectedClusterId: PropTypes.string,
  // TODO: remove the below eslint disable once routeName is used
  // eslint-disable-next-line react/no-unused-prop-types
  routeName: PropTypes.string.isRequired,
  handleSorting: PropTypes.func,
  sortOrder: PropTypes.string,
  clusters: PropTypes.array,
};

ClusterList.defaultProps = {
  dispatch: () => {},
  selectedClusterId: '',
  handleSorting: () => {},
  sortOrder: 'none',
  clusters: [],
};

const mapStateToProps = (state) => ({
  // Note:  controls whether the selectedCluster is highlighted or not.
  // if routeName is TOPIC_REVIEW, it is highlighted/expanded state
  // if routeName is Graph or Flow diagram - it is collapsed
  selectedClusterId: clusterSelectors.getSelectedClusterId(state) || '',
  routeName: getPathnameSelector(state),
});

const mapDispatchToProps = (dispatch) => ({
  dispatch,
});


export default connect(mapStateToProps, mapDispatchToProps)(ClusterList);
