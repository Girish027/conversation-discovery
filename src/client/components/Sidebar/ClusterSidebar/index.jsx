import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Popover from 'react-tiny-popover';
import _ from 'lodash';
import { List, Map, fromJS } from 'immutable';
import memoizeOne from 'memoize-one';
import FilterTopic from './FilterTopics';
import SearchBox from '../../SearchBox';
import '../../../styles/CollapsibleOuter.scss';
import {
  Filter,
  ContextualActionsBar,
  ContextualActionItem
} from '@tfs/ui-components';
import ClustersList from './ClusterList';
import {
  clusterSelectors,
  clusterActions,
  initialFilterState,
} from 'state/clusterState';
import { runSelectors } from 'state/runsState';

import Constants from 'components/../constants';
import { saveStorage } from '../../../utils/storageManager';
import { sidebarButtonStyle, popoverContainerStyle } from '../../../styles';
import constants from '../../../constants';
import { prepareClusterName } from '../../../utils/stringOperations';

const filterByFinalizeAndOpen = (cluster, filters) => {
  const initialFilters = initialFilterState;
  if (_.isEqual(filters, initialFilters)) {
    return true;
  }
  const isFinalized = cluster.finalized === filters.finalized;
  const isOpen = !cluster.finalized === filters.open;

  return isFinalized || isOpen;
};

const getMatchedClusters = (clusters, searchString) => {
  let clustersList = clusters.toJS();
  const search = searchString.trim().toLowerCase();

  if (search.length > 0) {
    clustersList = clustersList.filter((cluster) => cluster.rollupCluster.toLowerCase().includes(search)
        || prepareClusterName(cluster.clusterName).toLowerCase().includes(search)
    );
  }
  return clustersList;
};

function getFilteredClusters(clusters, count) {
  return [...clusters].sort((a, b) => (parseInt(a.count, 10) < parseInt(b.count, 10) ? 1 : -1)).splice(0, count);
}

function filterBySimilarityAndVolume(filteredClusters, filters, volume, clusters) {
  if (!_.volume) {
    if ({}.hasOwnProperty.call(volume, constants.filterClusterByVolume.volumeByCount) && volume.volumeByCount !== null) {
      filteredClusters = getFilteredClusters(clusters, volume.volumeByCount);
    } else if ({}.hasOwnProperty.call(volume, constants.filterClusterByVolume.volumeByPercent) && volume.volumeByPercent !== null) {
      const topClusters = Math.round([...clusters].length * volume.volumeByPercent / 100);
      filteredClusters = getFilteredClusters(clusters, topClusters);
    }
  }

  filteredClusters = filteredClusters.filter((cluster) => cluster.similarityCutOff >= filters.similarityCutOff);

  return filteredClusters;
}

const getGroupedClusters = memoizeOne((clustersImmutable = List([]), filtersImmutable = List([]), dispatch) => {
  const clusters = clustersImmutable;
  const filters = filtersImmutable.toJS();
  const newClusterList = [];

  const groupedCluster = [];
  let filteredClusters = [...clusters];

  const { volume } = filters;

  filteredClusters = filterBySimilarityAndVolume(filteredClusters, filters, volume, clusters);

  filteredClusters.forEach((val) => {
    const show = filterByFinalizeAndOpen(val, filters);
    if (show) {
      const hasCluster = groupedCluster.find((cluster) => cluster.rollupCluster === val.rollupCluster);
      if (hasCluster) {
        hasCluster.clusters.push(val);
        hasCluster.count += val.count;
      } else {
        groupedCluster.push({
          rollupCluster: val.rollupCluster,
          count: val.count,
          clusters: [val],
        });
      }
      newClusterList.push(val);
      dispatch(clusterActions.setFilteredClusters(newClusterList));
    }
  });
  return groupedCluster;
});

export class ClustersSidebar extends Component {
  constructor(props) {
    super(props);
    this.props = props;
    this.onSearch = this.onSearch.bind(this);
    this.onClickExpand = this.onClickExpand.bind(this);
    this.handleSorting = this.handleSorting.bind(this);
    this.onClickFilter = this.onClickFilter.bind(this);
    this.handleClosePopover = this.handleClosePopover.bind(this);
    this.getCollapbibleButton = this.getCollapbibleButton.bind(this);

    this.state = {
      clusters: [],
      searchString: '',
      sortOrder: Constants.sortOrder.none,
      isOpen: false,
      isPopoverOpen: false,
    };
  }

  componentDidMount() {
    const { dispatch } = this.props;
    dispatch(clusterActions.loadFilter());
    dispatch(clusterActions.getAllClusters());
  }

  static getDerivedStateFromProps(props, state) {
    const { clusters: existingClustersList, searchString } = state;
    const { clusters: newClusters, filters: newFilters, dispatch } = props;

    // This logic is for performing search operation on the clusters which returns filtered clusters(matched clusters)
    const newCluster = searchString ? getMatchedClusters(newClusters, searchString) : newClusters.toJS();
    const existingClusters = [];
    const newState = {};

    const initialFilters = initialFilterState;
    const filters = newFilters.toJS();

    // This logic is for comparing the existing clusters with new clusters(when props changes)
    existingClustersList.forEach((cluster) => {
      existingClusters.push(...cluster.clusters);
    });
    newCluster.sort((a, b) => a.clusterId - b.clusterId);
    existingClusters.sort((a, b) => a.clusterId - b.clusterId);

    if (!_.isEqual(existingClusters, newCluster) || !_.isEqual(filters, initialFilters)) {
      newState.clusters = getGroupedClusters(newCluster, newFilters, dispatch);
    }
    return newState;
  }

  onSearch(searchString) {
    this.setState({
      searchString
    });
  }

  onClickExpand() {
    this.setState({
      isOpen: !this.state.isOpen
    });
  }

  onClickFilter() {
    this.setState({
      isPopoverOpen: !this.state.isPopoverOpen
    });
  }

  getCollapbibleButton(isOpen) {
    let buttonClass = 'CollapsibleOuter';
    if (isOpen) {
      buttonClass = 'CollapsibleOuter isOpen';
    }

    return (
      <div>
        <span className={buttonClass}>
          {!isOpen
            ? Constants.buttons.expandAll
            : Constants.buttons.collapseAll
          }
        </span>
      </div>
    );
  }

  handleSorting() {
    const { sortOrder } = this.state;
    if (sortOrder !== Constants.sortOrder.asc) {
      this.setState({
        sortOrder: Constants.sortOrder.asc,
      });
      saveStorage(Constants.clusterSort, { sortOrder: Constants.sortOrder.asc });
    } else {
      this.setState({
        sortOrder: Constants.sortOrder.desc,
      });
      saveStorage(Constants.clusterSort, { sortOrder: Constants.sortOrder.desc });
    }
  }

  handleClosePopover() {
    this.setState({
      isPopoverOpen: false
    });
  }

  render() {
    const {
      isOpen, sortOrder, isPopoverOpen, clusters
    } = this.state;

    return (
      <div>
        <div style={{ display: 'flex' }}>
          <SearchBox
            onSearch={this.onSearch}
            placeholder='Find Intents'
            styleOverride={{ width: '90%' }}
          />

          <Popover
            isOpen={isPopoverOpen}
            position={'bottom'}
            align={'start'}
            onClickOutside={this.handleClosePopover}
            containerStyle={popoverContainerStyle}
            content={(
              <FilterTopic />
            )}
          >
            <ContextualActionsBar styleOverride={{ ...sidebarButtonStyle.containerStyle, width: '48%' }}>
              <ContextualActionItem
                right
                icon={Filter}
                onClickAction={this.onClickFilter}
                styleOverride={{ fontSize: '12px', marginRight: '-7px', marginTop: '23px' }}
              >
                FILTER
              </ContextualActionItem>
            </ContextualActionsBar>
          </Popover>
        </div>
        <ClustersList
          isOpen={isOpen}
          clusters={clusters}
          handleSorting={this.handleSorting}
          sortOrder={sortOrder}
        />
      </div>
    );
  }
}

ClustersSidebar.propTypes = {
  dispatch: PropTypes.func,
  // eslint-disable-next-line react/no-unused-prop-types
  filters: PropTypes.object,
};

ClustersSidebar.defaultProps = {
  dispatch: () => {},
  filters: fromJS(initialFilterState),
};

const mapStateToProps = (state) => {
  const allFilters = clusterSelectors.getFilters(state) || new Map({});
  const selectedRunId = runSelectors.getSelectedRunId(state) || '';
  const filters = allFilters.get(selectedRunId);
  return {
    clusters: clusterSelectors.getClusterList(state) || new List([]),
    filters,
  };
};

const mapDispatchToProps = (dispatch) => ({
  dispatch,
});


export default connect(mapStateToProps, mapDispatchToProps)(ClustersSidebar);
