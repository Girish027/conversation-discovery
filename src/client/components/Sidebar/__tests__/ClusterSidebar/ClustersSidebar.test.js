import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { fromJS } from 'immutable';
import { ClustersSidebar } from '../../ClusterSidebar';
import Constants from '../../../../constants';
import { saveStorage, loadStorage } from '../../../../utils/storageManager';

jest.useFakeTimers();


describe('<ClustersSidebar />', () => {
  let wrapper;

  const clusterState = {
    selectedClusterId: '3',
    clusters: [{
      clusterId: '1',
      clusterName: 'cancel_order_customer_service', // this is the one getting edited
      originalName: 'cancel_order_customer_service',
      clusterDescription: 'relates to cancel order',
      rollupCluster: 'Cancel_Order',
      suggestedNames: ['cancel_order_customer_service', 'abcde', 'defg'],
      count: 230,
      finalized: true,
      modifiedBy: 'user1@247.ai',
      modified: 1554930400001,
      finalizedBy: '',
    }, {
      clusterId: '2',
      clusterName: 'Cancel_Order',
      originalName: 'Cancel_Order',
      clusterDescription: 'order canceled',
      rollupCluster: 'Cancel_Order',
      suggestedNames: ['cancel_order_service', 'abcde', 'defg'],
      count: 400,
      finalized: true,
      modifiedBy: 'user1@247.ai',
      modified: 1554930400002,
      finalizedBy: '',
    }, {
      clusterId: '3',
      clusterName: 'store_close_holiday',
      originalName: 'store_close',
      clusterDescription: 'store closed due to reasons',
      rollupCluster: 'Store_Close',
      suggestedNames: ['merchant_close', 'abcde', 'defg'],
      count: 218,
      finalized: true,
      modifiedBy: 'user1@247.ai',
      modified: 1554930400003,
      finalizedBy: 'user1@247.ai',
    }
    ]
  };

  const props = {
    clusters: fromJS(clusterState.clusters),
  };


  afterAll(() => {
    jest.clearAllMocks();
  });

  const getShallowWrapper = (propsObj) => shallow(<ClustersSidebar
    {...propsObj}
  />);

  describe('Creating an instance', () => {
    test('should exist', () => {
      wrapper = getShallowWrapper(props);
      expect(wrapper.exists()).toBe(true);
    });
  });

  describe('Snapshots', () => {
    test('renders correctly', () => {
      wrapper = getShallowWrapper(props);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });


  describe('Functionality:', () => {
    afterEach(() => {
      jest.clearAllMocks();
    });

    describe('onSearch', () => {
      test('should change the state and update local storage when onSearch is called', () => {
        wrapper = getShallowWrapper(props);

        const searchString = 'Order cancel';
        wrapper.instance().onSearch(searchString);
        expect(wrapper.state().searchString).toEqual(searchString);

        saveStorage(Constants.clusterSearchString, searchString);
        expect(loadStorage(Constants.clusterSearchString)).toEqual(searchString);
      });
    });

    describe('onClickExpand', () => {
      test('should change the state when onClickExpand is called', () => {
        wrapper = getShallowWrapper(props);

        wrapper.instance().onClickExpand();
        expect(wrapper.state().isOpen).toEqual(true);

        wrapper.instance().onClickExpand();
        expect(wrapper.state().isOpen).toEqual(false);
      });
    });

    describe('onClickFilter', () => {
      test('should change the state of Popover', () => {
        wrapper = getShallowWrapper(props);

        wrapper.instance().onClickFilter();
        expect(wrapper.state().isPopoverOpen).toEqual(true);

        wrapper.instance().onClickFilter();
        expect(wrapper.state().isPopoverOpen).toEqual(false);
      });
    });

    describe('handleClosePopover', () => {
      test('should change the state of Popover to false', () => {
        wrapper = getShallowWrapper(props);

        wrapper.instance().onClickFilter();
        expect(wrapper.state().isPopoverOpen).toEqual(true);

        wrapper.instance().handleClosePopover();
        expect(wrapper.state().isPopoverOpen).toEqual(false);
      });
    });

    describe('handleSorting', () => {
      test('should change the state of sortOrder', () => {
        wrapper = getShallowWrapper(props);
        wrapper.instance().handleSorting();
        expect(wrapper.state().sortOrder).toEqual('asc');
        wrapper.instance().handleSorting();
        expect(wrapper.state().sortOrder).toEqual('desc');
      });
    });
  });
});
