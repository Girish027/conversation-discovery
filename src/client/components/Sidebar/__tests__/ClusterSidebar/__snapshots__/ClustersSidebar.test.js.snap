// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClustersSidebar /> Snapshots renders correctly 1`] = `
<div>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <SearchBox
      closeIcon={true}
      onSearch={[Function]}
      placeholder="Find Intents"
      styleOverride={
        Object {
          "width": "90%",
        }
      }
    />
    <Popover
      align="start"
      containerClassName="react-tiny-popover-container"
      containerStyle={
        Object {
          "backgroundColor": "white",
          "boxShadow": "0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)",
          "color": "black",
          "display": "flex",
          "fontSize": "13px",
          "fontWeight": 100,
          "height": "auto",
          "left": "300px",
          "padding": "25px",
          "userSelect": "none",
          "width": "auto",
        }
      }
      content={<Connect(FilterTopics) />}
      isOpen={false}
      onClickOutside={[Function]}
      padding={6}
      position="bottom"
      windowBorderPadding={6}
    >
      <withTheme(ContextualActionsBar)
        styleOverride={
          Object {
            "border": "unset",
            "fontWeight": "bold",
            "height": "45px",
            "width": "48%",
          }
        }
      >
        <withTheme(ContextualActionItem)
          icon={[Function]}
          onClickAction={[Function]}
          right={true}
          styleOverride={
            Object {
              "fontSize": "12px",
              "marginRight": "-7px",
              "marginTop": "23px",
            }
          }
        >
          FILTER
        </withTheme(ContextualActionItem)>
      </withTheme(ContextualActionsBar)>
    </Popover>
  </div>
  <Connect(ClusterList)
    clusters={Array []}
    handleSorting={[Function]}
    isOpen={false}
    sortOrder="none"
  />
</div>
`;
